File: results/0.5/ARP/arp-gpt-4o.c
#include <stdio.h>
#include <stdlib.h>
#include <hammer/hammer.h>

HParser *arp_parser() {
    HParser *htype = h_uint16();
    HParser *ptype = h_uint16();
    HParser *hlen = h_uint8();
    HParser *plen = h_uint8();
    HParser *oper = h_uint16();
    HParser *sha = h_length_value(hlen, h_uint8());
    HParser *spa = h_length_value(plen, h_uint8());
    HParser *tha = h_length_value(hlen, h_uint8());
    HParser *tpa = h_length_value(plen, h_uint8());

    return h_sequence(htype, ptype, hlen, plen, oper, sha, spa, tha, tpa, NULL);
}

void parse_arp(const char *filename) {
    FILE *file = fopen(filename, "rb");
    if (!file) {
        perror("Error opening file");
        exit(EXIT_FAILURE);
    }

    fseek(file, 0, SEEK_END);
    long file_size = ftell(file);
    fseek(file, 0, SEEK_SET);

    unsigned char *data = (unsigned char *)malloc(file_size);
    if (!data) {
        perror("Memory allocation failed");
        fclose(file);
        exit(EXIT_FAILURE);
    }

    fread(data, 1, file_size, file);
    fclose(file);

    HParser *parser = arp_parser();
    HParseResult *result = h_parse(parser, data, file_size);

    if (result) {
        printf("ARP packet parsed successfully.\n");
        h_parse_result_free(result);
    } else {
        printf("Failed to parse ARP packet.\n");
    }

    free(data);
    h_delete(parser);
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <binary_file>\n", argv[0]);
        return EXIT_FAILURE;
    }

    parse_arp(argv[1]);
    return EXIT_SUCCESS;
}

File: results/0.5/ARP/arp-gpt-4o.rs
use nom::bytes::complete::take;
use nom::combinator::map_res;
use nom::error::ErrorKind;
use nom::number::complete::{be_u16, u8};
use nom::IResult;
use std::env;
use std::fs::File;
use std::io::{self, Read};

// ARP Packet structure
#[derive(Debug)]
struct ArpPacket {
    htype: u16,
    ptype: u16,
    hlen: u8,
    plen: u8,
    oper: u16,
    sha: Vec<u8>,
    spa: Vec<u8>,
    tha: Vec<u8>,
    tpa: Vec<u8>,
}

// Parse ARP packet
fn parse_arp(input: &[u8]) -> IResult<&[u8], ArpPacket> {
    let (input, htype) = be_u16(input)?;
    let (input, ptype) = be_u16(input)?;
    let (input, hlen) = u8(input)?;
    let (input, plen) = u8(input)?;
    let (input, oper) = be_u16(input)?;
    let (input, sha) = take(hlen)(input)?;
    let (input, spa) = take(plen)(input)?;
    let (input, tha) = take(hlen)(input)?;
    let (input, tpa) = take(plen)(input)?;

    Ok((
        input,
        ArpPacket {
            htype,
            ptype,
            hlen,
            plen,
            oper,
            sha: sha.to_vec(),
            spa: spa.to_vec(),
            tha: tha.to_vec(),
            tpa: tpa.to_vec(),
        },
    ))
}

fn main() -> io::Result<()> {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: {} <binary_file>", args[0]);
        return Ok(());
    }

    let mut file = File::open(&args[1])?;
    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer)?;

    match parse_arp(&buffer) {
        Ok((_, arp_packet)) => {
            println!("{:?}", arp_packet);
        }
        Err(err) => {
            eprintln!("Failed to parse ARP packet: {:?}", err);
        }
    }

    Ok(())
}